File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/push/exceptions/NonNormalizedPhoneNumberException.java
package org.whispersystems.signalservice.api.push.exceptions;

import com.fasterxml.jackson.annotation.JsonProperty;

import org.whispersystems.signalservice.internal.util.JsonUtil;

import io.reactivex.rxjava3.annotations.NonNull;

/**
 * Response indicating we gave the server a non-normalized phone number. The expected normalized version of the number is provided.
 */
public class NonNormalizedPhoneNumberException extends NonSuccessfulResponseCodeException {

  private final String originalNumber;
  private final String normalizedNumber;

  public static NonNormalizedPhoneNumberException forResponse(@NonNull String responseBody) throws MalformedResponseException {
    JsonResponse response = JsonUtil.fromJsonResponse(responseBody, JsonResponse.class);
    return new NonNormalizedPhoneNumberException(response.originalNumber, response.normalizedNumber);
  }

  public NonNormalizedPhoneNumberException(String originalNumber, String normalizedNumber) {
    super(400);

    this.originalNumber   = originalNumber;
    this.normalizedNumber = normalizedNumber;
  }

  public String getOriginalNumber() {
    return originalNumber;
  }

  public String getNormalizedNumber() {
    return normalizedNumber;
  }

  private static class JsonResponse {
    @JsonProperty
    private String originalNumber;

    @JsonProperty
    private String normalizedNumber;
  }
}

File: lib/libsignal-service/src/main/java/org/whispersystems/signalservice/api/push/exceptions/NonNormalizedPhoneNumberException.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.whispersystems.signalservice.api.push.exceptions

import org.signal.network.exceptions.MalformedResponseException
import org.signal.network.exceptions.NonSuccessfulResponseCodeException
import org.signal.network.util.JsonUtil

/**
 * Response indicating we gave the server a non-normalized phone number. The expected normalized version of the number is provided.
 */
class NonNormalizedPhoneNumberException(
  val originalNumber: String,
  val normalizedNumber: String
) : NonSuccessfulResponseCodeException(400) {

  /** The 400 response body. Separate from the exception itself so that it can be deserialized. */
  class Body(
    val originalNumber: String? = null,
    val normalizedNumber: String? = null
  )

  companion object {
    @JvmStatic
    @Throws(MalformedResponseException::class)
    fun forResponse(responseBody: String): NonNormalizedPhoneNumberException {
      val body = JsonUtil.fromJsonResponse(responseBody, Body::class.java)

      if (body.originalNumber == null || body.normalizedNumber == null) {
        throw MalformedResponseException("Response is missing a number")
      }

      return NonNormalizedPhoneNumberException(body.originalNumber, body.normalizedNumber)
    }
  }
}
